[BEAM-6394] Add support to write protobuf data using ProtoParquetReader

cyberbeam524 opened the pull request:
**Please** add a meaningful description for your change here

- fixes #19366 
- better handling of parquet‑protobuf data by switching to ProtoParquetReader, and clearer configuration via display data

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

cyberbeam524 commented:
assign set of reviewers

cyberbeam524 commented:
@Abacn Could you please take a look?

Abacn commented on sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java:
```diff
@@ -746,6 +780,31 @@ public void processElement(
+
+          // Get the configuration and set the property using the literal.
+          Configuration conf2 = getConfWithModelClass();
+          conf2.set("parquet.proto.ignore.unknown.fields", "TRUE");
```
What is the purpose of "ignore.unknown.fields" here? In general we should not set unsafe configure for user if they are not aware of

Abacn commented on sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java:
```diff
@@ -746,6 +780,31 @@ public void processElement(
+            }
+          }
+
+          return; // exit after using the proto path
```
For readability, I would recommend structure the code, e.g. 

getParquetFileReader() {
  switch (FORMAT) {
    case AVRO: return getParquetAvroReader()
    case PROTO: return getParquetProtoReader()
  }
}
 instead of branching in place

Abacn reviewed (commented):
Thanks! This is a good start. Left a few comments

Abacn commented:
waiting on author

cyberbeam524 commented:
@Abacn Added the following changes according to your comments. Could you please take a look? Thank you!
- Added ReaderFormat enum to ParquetIO.ReadFiles
- Replaced boolean `useProtoReader` flag with a public `ReaderFormat` enum (AVRO, PROTO)
- Default to `ReaderFormat.AVRO` for backward compatibility
- Added `withAvroReader()` and `withProtoReader()` builder methods
- Refactored`SplitReadFn` to `switch` on `ReaderFormat` and dispatch to Avro or Proto reader
- Removed`"parquet.proto.ignore.unknown.fields"` setting from the default path
- Update DisplayData to emit `readerFormat.name()`
- Update unit tests to assert on `ReaderFormat` and display data accordingly

damccorm commented:
@Abacn could you please take another look?

derrickaw commented:
Hi @ahmedabu98 and @Abacn, please review when you have a chance.  Thanks!

damccorm commented:
R: @Abacn 

Keeping this sticky
